Improve Streamer Update Performance / Implementation

Ayiga opened the pull request:
<!-- Closes #<ISSUE_NUMBER> -->
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->
- Refactors the `Update` method of the `EspressoStreamer` to be smaller
- Updates the behavior of the `Update` method of `EspressoStreamer` to attempt to ensure that `Next` will return a `Batch` if able
- Makes the streamer_test.go tests pass with the implementation

This PR focuses on trying to make progress with the `EsperssoStreamer` by adding in some checks to the `Update` method to keep processing more HotShot blocks if there still isn't a new `Batch` available in the `EspressoStreamer`.

The idea here is that the call to `Update` should do it's best to ensure that the next call to `Next` should return a `Batch` if it is able to do so.  This seems to be a reasonable expectation for how a user would consume the `Streamer`.

Additionally, it keeps advancing the `hotShotPos` upon successfully retrieving the target block from HotShot.

It is essentially to keep advancing this whenever we successfully retrieve a HotShot block, as not doing so will result in us retrieving the same HotShot `Block`s over and over again without progress, unnecessarily. 

<!-- ### This PR does not: -->
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

<!-- ### Key places to review: -->
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210210306238898

Ayiga committed:
Split `Update` method in EspressoStreamer

The `Update` method is doing a lot of things within a single closure
that can make it difficult to track / reason about.  Some of these
concepts can be better isolated to help reduce cognitive load of
reading this implementation, while still persisting the underlying
concepts of what each section is achieving.

This change refactors the `Update` method to move some of these
more isolate concepts together while preserving the original
implementation's behavior.  This is just an organizational change,
and not a functional one.

Ayiga committed:
Update `Update` method to try and ensure batch availability

The `EspressoStreamer` was written with the idea to load blocks
from `HotShot` in discrete batches.  This is a great idea, but it may
break standard consumer assumptions about how streams tend to
work.

In general, it doesn't hold the principal that calling `Update`
will load **AT LEAST** the next available `Batch` into the `Streamer`
or fail due to an unforeseen error or having no actual next `Batch`
be currently available.  I believe this is a reasonable expectation that
`Update` should do it's best to fulfill, as not doing this will just result
in the user calling `Update` and `Next` again to get to this desired
result anyway.

This change refactors `Update` some more, and allows it to process
more Batches to try and ensure that there is a next `Batch` loaded
into the `BatchBuffer`.

philippecamacho commented:
@Ayiga Artemii PR #119 has been merged. Would you like to merge celo-integration back into this branch (there is a conflict for `espresso/streamer.go`?

philippecamacho commented on espresso/streamer.go:
```diff
@@ -113,6 +113,15 @@ func NewEspressoStreamer[B Batch](
+	// TODO: This will need to be updated to fallback to a known good starting
+	//       position in the future. This is a temporary solution to ensure that
+	//       we will, at the very least, not stall and fail to make progress.
+	s.hotShotPos = 0
```
Why not keep `s.hotShotPos = s.fallbackHotShotPos`. After Artemii's [update](https://github.com/EspressoSystems/optimism-espresso-integration/pull/119/files#diff-1912bc32db1ca7d02cb1784e7e053dc6d5a6e91043028c8160d48afaf23d2d3fR341-R360) isn't s.fallbackHotShotPos set correctly?

philippecamacho replied:
Oh, this will be possible once you merge with celo-integration.

Ayiga replied:
Rebased from `celo-integration` and force pushed at 0c5e61e913eedd4ec62dc6631f9262e29cb9464d.

philippecamacho commented on espresso/streamer.go:
```diff
@@ -193,40 +193,125 @@ func (s *EspressoStreamer[B]) CheckBatch(ctx context.Context, batch B) (BatchVal
+	i := start
 
+	for i <= currentBlockHeight {
+		s.Log.Info("Remaining list before", "Size", len(s.RemainingBatches))
```
Let us remove this log message which was for debugging purposes.

Ayiga replied:
Removed in 887e269f8593775c3b85e7052cda5310703361ec.

philippecamacho commented on espresso/streamer.go:
```diff
@@ -193,40 +193,125 @@ func (s *EspressoStreamer[B]) CheckBatch(ctx context.Context, batch B) (BatchVal
+		// Process the remaining batches
+		s.processRemainingBatches(ctx)
+
+		s.Log.Info("Remaining list after", "Size", len(s.RemainingBatches))
```
Let us remove this log message which was for debugging purposes.

Ayiga replied:
Removed in 887e269f8593775c3b85e7052cda5310703361ec.

philippecamacho commented on espresso/streamer.go:
```diff
@@ -193,40 +193,125 @@ func (s *EspressoStreamer[B]) CheckBatch(ctx context.Context, batch B) (BatchVal
+		}
+		// Our position has advanced to the block following the finish block
+		// we just passed.
+		i = finish + 1
```
I have the feeling that in some cases we might end up skipping some Espresso blocks.
Why not just increment `i` by 1 and have it range through start...finish?

philippecamacho replied:
Oh I guess I am confusing this loop with what happens in `processHotShotRange`.

Ayiga replied:
Right, the range is being processed in the `processHotShotRange`.  Provided that no error is returned from calling that method, it will process the given `start` to `finish` **INCLUSIVELY**, which is why I'm setting `i` to `finish` + 1 (It would always need to be set to at least `finish`). It's a small optimization, the increment **doesn't** actually need to happen, but doing it will allow us to not download the `finish` block again.  This loop here is tracking the position, just to ensure that we don't exceed the `currentBlockHeight` we've already pulled.

philippecamacho replied:
I understand now thanks.

Ayiga committed:
Remove logging statements used for debugging

philippecamacho reviewed (approved):
LGTM

QuentinI commented on espresso/streamer.go:
```diff
@@ -193,40 +193,123 @@ func (s *EspressoStreamer[B]) CheckBatch(ctx context.Context, batch B) (BatchVal
 	s.Log.Info("Fetching hotshot blocks", "from", start, "upTo", finish)
+	i := start
 
+	for i <= currentBlockHeight {
```
Nit: I feel like this loop would read smoother if rephrased solely in terms of start and finish, i.e.
```go
for {
	start, finish := s.computeEspressoBlockHeightsRange(currentBlockHeight)
	if start >= finish {
	   break
	}
	// ....
}
```
Not going to die on this hill though, feel free to resolve if you disagree

Ayiga replied:
I think this is fair.  I'll modify the local `i` variable and just use `start` instead.

Ayiga replied:
Addressed in c269c0d8f18151be6c0ae77c37e2cb30b4033729.  This introduced an edge case that was then fixed in 4b3f7f48270aa4c3172344ef21e51257a78b71eb.

QuentinI commented on espresso/streamer.go:
```diff
@@ -251,71 +334,51 @@ func (s *EspressoStreamer[B]) Update(ctx context.Context) error {
 		}
 
-		for _, transaction := range txns.Transactions {
+		s.Log.Info("Inserting batch into buffer", "batch", batch)
```
Nit: we aren't yet inserting here, in fact we have exact same message on the penultimate line of the function, which could be confusing when grepping through logs.

Ayiga replied:
Removed in 939b3f991fbdfa08810f29e2ddb9252c96f3c4b3.

QuentinI reviewed (approved):
This is a significant improvement to readability! Nitpicked a little, but overall this is great

Ayiga committed:
Remove unnecessary log

Ayiga committed:
Refactor local `Update` variables for clarity

Based on feedback provided by @QuentinI, continually utilizing the
variables `start`, and `finish` will provide better clarity over using a
local increment variable `i`:
https://github.com/EspressoSystems/optimism-espresso-integration/pull/129/files#r2089587647

The `i` variables instead of `processRange` has been renamed to be
`height` to better signify it's intention.

The loop inside of `Update` has been reworked to exclusively rely on
`start` and `finish` instead of using a local increment `i` variable.

The `start` position calculated in `computeEspressoBlockHeightsRange`
has been modified to start at the next block, instead of the the
`hotShotPos`, as the `hotShotPos` indicates a block we've already
processed (provided it is not the starting block)

Ayiga committed:
Fix an Espresso Block Processing edge case

There is an edge case where the initial block being processed from
espresso actually contains the next batch.  We need to process the initial
block, so we need to loosen the restrictions on the condition:
`start` >= `finish`.  However, when we loosen this restriction we end up
with the potential to get stuck in an infinite loop if the only block available
from Espresso is its initial block.

We add some safe guards to prevent this edge case while still ensuring
that it processes the initial block at least once.
